README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
testlib
=======

[![Travis-CI Build Status](https://travis-ci.org/rsquaredacademy/testlib.svg?branch=master)](https://travis-ci.org/rsquaredacademy/testlib) [![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/rsquaredacademy/testlib?branch=master&svg=true)](https://ci.appveyor.com/project/rsquaredacademy/testlib) [![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/testlib)](http://cran.r-project.org/package=testlib) [![Coverage Status](https://img.shields.io/codecov/c/github/rsquaredacademy/testlib/master.svg)](https://codecov.io/github/rsquaredacademy/testlib?branch=master) [![Project Status: Active - The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/0.1.0/active.svg)](http://www.repostatus.org/#active) [![GitHub issues](https://img.shields.io/github/issues/rsquaredacademy/testlib.svg)](https://github.com/rsquaredacademy/testlib/issues) [![Twitter](https://img.shields.io/twitter/url/https/github.com/rsquaredacademy/testlib.svg?style=social)](https://twitter.com/intent/tweet?text=Wow:&url=%5Bobject%20Object%5D)

Overview
--------

testlib is a demo R package created to explore the different aspects of R package development. Hadley Wickham's book [R Packages](http://r-pkgs.had.co.nz/) was used as a guide for developing this package. The package contains a function for creating two way tables of categorical data and visualizing such tables using bar and mosaic plots. Learn more in `vignette("testlib")` or `?cross_table`.

Installation
------------

testlib is not currently available from CRAN, but you can install the development version from github:

[![GitHub version](https://badge.fury.io/gh/rsquaredacademy%2Ftestlib.svg)](https://badge.fury.io/gh/rsquaredacademy%2Ftestlib)

``` r
# install.packages("devtools")
devtools::install_github("rsquaredacademy/testlib")
```

Usage
-----

The main function in the pacakge is `cross_table()` which creates two way tables of categorical data.

``` r
library(testlib)
k <- testlib::cross_table(mtcars$cyl, mtcars$am)
k
```

    ##     Cell Contents
    ##  |---------------|
    ##  |     Frequency |
    ##  |       Percent |
    ##  |       Row Pct |
    ##  |       Col Pct |
    ##  |---------------|
    ## 
    ##  Total Observations:  32 
    ## 
    ## -------------------------------------------------------------
    ## |              |                     am                     |
    ## -------------------------------------------------------------
    ## |          cyl |            0 |            1 |    Row Total |
    ## -------------------------------------------------------------
    ## |            4 |            3 |            8 |           11 |
    ## |              |        0.094 |         0.25 |              |
    ## |              |         0.27 |         0.73 |         0.34 |
    ## |              |         0.16 |         0.62 |              |
    ## -------------------------------------------------------------
    ## |            6 |            4 |            3 |            7 |
    ## |              |        0.125 |        0.094 |              |
    ## |              |         0.57 |         0.43 |         0.22 |
    ## |              |         0.21 |         0.23 |              |
    ## -------------------------------------------------------------
    ## |            8 |           12 |            2 |           14 |
    ## |              |        0.375 |        0.062 |              |
    ## |              |         0.86 |         0.14 |         0.44 |
    ## |              |         0.63 |         0.15 |              |
    ## -------------------------------------------------------------
    ## | Column Total |           19 |           13 |           32 |
    ## |              |        0.594 |        0.406 |              |
    ## -------------------------------------------------------------

License
-------

[![License](http://img.shields.io/:license-mit-blue.svg)](https://github.com/rsquaredacademy/testlib/blob/master/LICENSE)


NEWS.md
# testlib 0.1.0

* First release



tests/testthat.R
library(testthat)
library(testlib)

test_check("testlib")


tests/testthat/test-utils.R
context('utils')

test_that('output from l matches the expected result', {

    expect_equal(l(deparse(substitute(mtcars$mpg))), "mpg")
    expect_equal(l(deparse(substitute(mpg))), "mpg")
    expect_equal(l(deparse(substitute(mtcars@mpg))), "mtcars@mpg")

})

test_that('output from row_pct matches the expected result', {

    m <- matrix(c(12, 23, 37, 21), nrow = 2)
    exptd <- matrix(c(0.245, 0.523, 0.755, 0.477), nrow = 2)
    expect_equivalent(round(row_pct(m, c(49, 44)), 3), exptd)

})


test_that('output from col_pct matches the expected result', {

    m <- matrix(c(12, 23, 37, 21), nrow = 2)
    exptd <- matrix(c(0.343, 0.657, 0.638, 0.362), nrow = 2)
    expect_equivalent(round(col_pct(m, c(35, 58)), 3), exptd)

})


test_that('output from formatter matches the expected result', {

    expect_equal(formatter(3), "            3")

})


tests/testthat/test-output.R
context('print_cross')

test_that('output from print_cross matches expected output', {

    k <- cross_table(mtcars$cyl, mtcars$vs)
    pim <- "    Cell Contents
|---------------|
|     Frequency |
|       Percent |
|       Row Pct |
|       Col Pct |
|---------------|
Total Observations:  32
-------------------------------------------------------------
|              |                     vs                     |
-------------------------------------------------------------
|          cyl |            0 |            1 |    Row Total |
-------------------------------------------------------------
|            4 |            1 |           10 |           11 |
|              |        0.031 |        0.312 |              |
|              |         0.09 |         0.91 |         0.34 |
|              |         0.06 |         0.71 |              |
-------------------------------------------------------------
|            6 |            3 |            4 |            7 |
|              |        0.094 |        0.125 |              |
|              |         0.43 |         0.57 |         0.22 |
|              |         0.17 |         0.29 |              |
-------------------------------------------------------------
|            8 |           14 |            0 |           14 |
|              |        0.438 |            0 |              |
|              |            1 |            0 |         0.44 |
|              |         0.78 |            0 |              |
-------------------------------------------------------------
| Column Total |           18 |           14 |           32 |
|              |        0.563 |        0.437 |              |
-------------------------------------------------------------"

    expect_output(print_cross(k), pim)
})


R/utils.R
rounda <- function(x) {
    round(x, 2)
}

formatter <- function(x) {
    format(as.character(x), width = 13, justify = "right")
}

row_pct <- function(mat, tot) {
    d <- dim(mat)
    rows <- d[1]
    l <- length(tot)
    result <- c()
    for (i in seq_len(rows)) {
        diva <- mat[i, ]/tot[i]
        result <- rbind(result, diva)
    }
    rownames(result) <- NULL
    return(result)
}

col_pct <- function(mat, tot) {
    d <- dim(mat)
    cols <- d[2]
    l <- length(tot)
    result <- c()
    for (i in seq_len(cols)) {
        diva <- mat[, i]/tot[i]
        result <- cbind(result, diva)
    }
    colnames(result) <- NULL
    return(result)
}

l <- function(x) {
    x <- as.character(x)
    k <- grep("\\$", x)
    if (length(k) == 1) {
        temp <- strsplit(x, "\\$")
        out <- temp[[1]][2]
    } else {
        out <- x
    }
    return(out)
}



R/cross-table.R
#' @title Two Way Tables
#' @description \code{cross_table} creates two way tables of categorical
#' variables. The tables created can be visualized as barplots and mosaicplots.
#'
#' @param var1 First categorical variable
#' @param var2 Second categorical variable
#' @details cross_table is a generic function
#' @return \code{cross_table} returns an object of class \code{"cross_table"}.
#' An object of class \code{"cross_table"} is a list containing at least the
#' following components:
#'
#' \item{obs}{obs number of observations}
#' \item{var2_levels}{levels of the second categorical variable}
#' \item{var1_levels}{levels of the first categorical variable}
#' \item{varnames}{names of the variables}
#' \item{twowaytable}{table of the variables}
#' \item{percent_table}{table of percentages}
#' \item{row_percent}{table of row percentages}
#' \item{col_percent}{table of column percentages}
#' \item{column_totals}{total of columns}
#' \item{percent_column}{total of columns as a percentage}
#'
#' @examples
#' k <- cross_table(mtcars$cyl, mtcars$am)
#'
#' # bar plots
#' plot(k)
#' plot(k, beside = TRUE)
#' plot(k, proportional = TRUE)
#'
#' # mosaic plots
#' mosaicplot(k)
#' @export

cross_table <- function(var1, var2) UseMethod("cross_table")

#' @export
#' @rdname cross_table
cross_table.default <- function(var1, var2) {

    var_1 <- l(deparse(substitute(var1)))
    var_2 <- l(deparse(substitute(var2)))
    var_names <- c(var_1, var_2)
    x <- as.matrix(table(var1, var2))
    rownames(x) <- NULL
    n <- length(var1)
    if (is.factor(var1)) {
        row_name <- levels(var1)
    } else {
        row_name <- unique(sort(var1))
    }
    per_mat <- round(x/n, 3)
    row_pct <- apply(per_mat, 1, sum)
    col_pct <- apply(per_mat, 2, sum)
    per_mat <- cbind(per_mat, row_pct)
    per_mat <- suppressWarnings(rbind(per_mat, col_pct))
    d <- dim(per_mat)
    per_mat[d[1], d[2]] <- 1
    rowtotal <- apply(x, 1, sum)
    coltotal <- apply(x, 2, sum)
    rcent <- row_pct(x, rowtotal)
    rcent <- cbind(rcent, row_pct)
    rcent <- apply(rcent, c(1, 2), rounda)
    ccent <- col_pct(x, coltotal)
    ccent <- apply(ccent, c(1, 2), rounda)
    x <- cbind(x, rowtotal)
    x <- cbind(unname(row_name), x)
    if (is.factor(var2)) {
        col_name <- levels(var2)
    } else {
        col_name <- unique(sort(var2))
    }

    result <- list(obs = n, var2_levels = col_name, var1_levels = row_name, varnames = var_names,
        twowaytable = x, percent_table = per_mat, row_percent = rcent, column_percent = ccent,
        column_totals = coltotal, percent_column = col_pct)


    class(result) <- "cross_table"
    return(result)
}

#' @export
print.cross_table <- function(x, ...) {

    print_cross(x)

}

#' @importFrom graphics barplot
#' @importFrom grDevices rainbow
#' @export
#' @title Cross Table Bar Plot
#' @description \code{barplot.cross_table} creates stacked and grouped bar plots
#' for the two way tables created using \code{cross_table}
#' @details Bar plot method added to cross_table
#' @param x An object of class cross_table
#' @param beside a logical value. If FALSE, the columns of height are portrayed
#' as stacked bars, and if TRUE the columns are portrayed as juxtaposed bars.
#' @param proportional a logical value. If TRUE, the height of the bars is
#' proportional
#' @param ... further arguments to be passed to or from methods.
#'
#' @examples
#' k <- cross_table(mtcars$cyl, mtcars$am)
#' plot(k)
#' plot(k, beside = TRUE)
#' plot(k, proportional = TRUE)
#'
plot.cross_table <- function(x, beside = FALSE, proportional = FALSE, ...) {
    i_data <- x$twowaytable
    nb <- ncol(i_data)
    bdata <- i_data[, c(-1, -nb)]
    ln <- length(x$var2_levels)
    bardata <- matrix(as.numeric(bdata), ncol = ln)
    cols <- nrow(bardata)


    # proportional stacked bar plots
    if (proportional == TRUE) {

        colbar <- colSums(bardata)
        nh <- nrow(bardata)
        h <- rep(colbar, nh)
        hichka <- matrix(h, nrow = nh, byrow = T)
        propo_data <- round((bardata/hichka) * 100, 2)
        barplot(propo_data, col = rainbow(cols), main = paste(x$varnames[1],
            "by", x$varnames[2]), xlab = x$varnames[2], ylab = x$varnames[1],
            legend.text = T)
        result <- list(data = propo_data)

    } else {

        barplot(bardata, col = rainbow(cols), beside = beside,
                main = paste(x$varnames[1], "by", x$varnames[2]),
                xlab = x$varnames[2], ylab = x$varnames[1],
                legend.text = T)
        result <- list(data = bardata)
    }

    class(result) <- "cross_table"
    return(result)
}

#' @importFrom graphics mosaicplot
#' @importFrom grDevices rainbow
#' @export
#' @title Mosaic Plot Cross Table
#' @description \code{mosaicplot.cross_table} creates mosaic plot
#' for the two way tables created using \code{cross_table}
#' @details Mosaic plot method added to cross_table
#' @param x An object of class cross_table
#' @param ... further arguments to be passed to or from methods.
#' @examples
#' k <- cross_table(mtcars$cyl, mtcars$am)
#' mosaicplot(k)
#'
mosaicplot.cross_table <- function(x, ...) {
    i_data <- x$twowaytable
    nb <- ncol(i_data)
    mdata <- i_data[, c(-1, -nb)]
    ln <- length(x$var2_levels)
    modata <- matrix(as.numeric(mdata), ncol = ln)
    cols <- nrow(modata)
    mosaicplot(modata, col = rainbow(cols), xlab = x$varnames[1], ylab = x$varnames[2],
        main = paste(x$varnames[1], "by", x$varnames[2]))
    result <- list(data = modata)
    class(result) <- "cross_table"
    return(result)
}


R/output.R
print_stats <- function(data) {
    
    n <- nchar(format(data$Uncorrect_SS, nsmall = 2))
    width1 <- 52 + (2 * n)
    width2 <- as.integer(width1/2)
    width3 <- width2 - 5
    width4 <- width2 - 2
    
    col1 <- max(nchar(as.character(data$Lowest_Obs)))
    col2 <- max(nchar(as.character(data$Highest_Obs)))
    col3 <- max(nchar(as.character(data$Lowest_Obs_Index)))
    col4 <- max(nchar(as.character(data$Highest_Obs_Index)))
    v <- nchar("Value")
    ol <- max(col1, col2, col3, col4, v)
    gap <- width4 - (2 * ol)
    
    cat(formatc("Univariate Analysis", width1), "\n\n", formatl("N"), formatr(data$Observations, 
        n), formats(), formatl("Variance"), formatr(data$Variance, n), "\n", formatl("Missing"), 
        formatr(data$Missing_Obs, n), formats(), formatl("Std Deviation"), formatr(data$Standard_Deviation, 
            n), "\n", formatl("Mean"), formatr(data$Mean, n), formats(), formatl("Range"), 
        formatr(data$Range, n), "\n", formatl("Median"), formatr(data$Median, n), 
        formats(), formatl("Interquartile Range"), formatr(data$Interquartile_Range, 
            n), "\n", formatl("Mode"), formatr(data$Mode, n), formats(), formatl("Uncorrected SS"), 
        formatr(data$Uncorrect_SS, n), "\n", formatl("Trimmed Mean"), formatr(data$Trimmed_mean, 
            n), formats(), formatl("Corrected SS"), formatr(data$Correct_SS, n), 
        "\n", formatl("Skewness"), formatr(data$Skewness, n), formats(), formatl("Coeff Variation"), 
        formatr(data$Coefficient_Of_Variation, n), "\n", formatl("Kurtosis"), formatr(data$Kurtosis, 
            n), formats(), formatl("Std Error Mean"), formatr(data$Std_Error_Mean, 
            n), "\n\n", formatc("Quantiles", width1), "\n\n", formatc("Quantile", 
            width2), formatc("Estimate", width2), "\n\n", formatc("100% Max  ", 
            width2), formatnc(data$Max, width2), "\n", formatc("99%       ", width2), 
        formatnc(data$Percentile_99, width2), "\n", formatc("95%       ", width2), 
        formatnc(data$Percentile_95, width2), "\n", formatc("90%       ", width2), 
        formatnc(data$Percentile_90, width2), "\n", formatc("75% Q3    ", width2), 
        formatnc(data$Percentile_75, width2), "\n", formatc("50% Median", width2), 
        formatnc(data$Median, width2), "\n", formatc("25% Q1    ", width2), formatnc(data$Percentile_25, 
            width2), "\n", formatc("10%       ", width2), formatnc(data$Percentile_10, 
            width2), "\n", formatc("5%        ", width2), formatnc(data$Percentile_05, 
            width2), "\n", formatc("1%        ", width2), formatnc(data$Percentile_01, 
            width2), "\n", formatc("0% Min    ", width2), formatnc(data$Min, width2), 
        "\n\n", formatc("Extreme Observations", width1), "\n\n", formatc("Lowest", 
            width2), formatc("Highest", width2), "\n\n", formatol("Value", ol), 
        format_gap(gap), formatol("Obs", ol), formats(), formatol("Value", ol), 
        format_gap(gap), formatol("Obs", ol), "\n")
    for (i in seq_len(5)) {
        cat("", formatol(data$Lowest_Obs[i], ol), format_gap(gap), formatol(data$Lowest_Obs_Index[i], 
            ol), formats(), formatol(data$Highest_Obs[i], ol), format_gap(gap), 
            formatol(data$Highest_Obs_Index[i], ol), "\n")
    }
    
}


print_fcont <- function(data) {
    
    cat(format(paste("Variable:", data$varname), width = 77, justify = "centre"), 
        "\n")
    cat("|---------------------------------------------------------------------------|
|                                 Cumulative                    Cumulative  |
|     Bins      |  Frequency   |   Frequency  |   Percent    |    Percent   |
|---------------------------------------------------------------------------|")
    for (i in seq_len(data$bins)) {
        k <- i + 1
        cat("\n|", formata(data$breaks[i], 1, 5), "-", formata(data$breaks[k], 
            1, 5), "|", formata(data$frequency[i], 2, 12), "|", formata(data$cumulative[i], 
            2, 12), "|", formatas(data$percent[i], 2, 12), "|", formatas(data$cum_percent[i], 
            2, 12), "|")
        cat("\n|---------------------------------------------------------------------------|")
    }
    
}


print_ftable <- function(data) {
    
    nr <- nrow(data$ftable)
    nc <- ncol(data$ftable)
    cat(format(paste("Variable:", data$varname), width = 76, justify = "centre"), 
        "\n")
    cat("|--------------------------------------------------------------------------|
|                                Cumulative                    Cumulative  |
|    Levels    |  Frequency   |   Frequency  |   Percent    |    Percent   |
|--------------------------------------------------------------------------|\n")
    for (i in seq_len(nr)) {
        for (j in seq_len(nc)) {
            cat("|", formatter_freq(data$ftable[i, j]))
        }
        cat("|")
        cat("\n|--------------------------------------------------------------------------|\n")
    }
    cat("\n\n")
    
}


print_cross <- function(data) {
    
    p <- length(data$var2_levels)
    q <- p + 2
    h <- p + 1
    r <- (h * 15) - 3
    f <- length(data$var1_levels)
    g <- f + 2
    h <- p + 1
    
    col_names <- c(data$varnames[1], data$var2_levels, "Row Total")
    col_totals <- c("Column Total", data$column_totals, data$obs)
    
    cat(formatter("    Cell Contents\n"), "|---------------|\n", "|", formatter("Frequency"), 
        "|\n", "|", formatter("Percent"), "|\n", "|", formatter("Row Pct"), "|\n", 
        "|", formatter("Col Pct"), "|\n", "|---------------|\n\n", "Total Observations: ", 
        data$obs, "\n\n")
    cat("-", rep("---------------", q), sep = "")
    cat("\n")
    cat("|              |", format(data$varnames[2], width = r, justify = "centre"), 
        "|")
    cat("\n")
    cat("-", rep("---------------", q), sep = "")
    cat("\n|")
    for (i in seq_along(col_names)) {
        cat(formatter(col_names[i]), "|")
    }
    cat("\n-", rep("---------------", q), sep = "")
    cat("\n")
    
    for (i in seq_len(f)) {
        cat("|")
        for (j in seq_len(q)) {
            cat(formatter(data$twowaytable[i, j]), "|")
        }
        cat("\n")
        cat("|              |")
        for (j in seq_len(p)) {
            cat(formatter(data$percent_table[i, j]), "|")
        }
        cat("              |")
        cat("\n")
        cat("|              |")
        for (j in seq_len(h)) {
            cat(formatter(data$row_percent[i, j]), "|")
        }
        cat("\n")
        cat("|              |")
        for (j in seq_len(p)) {
            cat(formatter(data$column_percent[i, j]), "|")
        }
        cat("              |")
        cat("\n-", rep("---------------", q), sep = "")
        cat("\n")
    }
    cat("|")
    for (i in seq_along(col_totals)) {
        cat(formatter(col_totals[i]), "|")
    }
    cat("\n")
    cat("|              |")
    for (i in seq_along(data$percent_column)) {
        cat(formatter(data$percent_column[i]), "|")
    }
    cat("              |")
    cat("\n-", rep("---------------", q), sep = "")
    cat("\n")
    
}


print_group <- function(data) {
    
    line <- 23
    n <- 21
    n_names <- max(nchar(data$stats[2, c(-1)]))
    n_uss <- max(nchar(data$stats[12, c(-1)]))
    w <- max(n_names, n_uss) + 2
    cola <- ncol(data$stats)
    col <- cola - 1
    ow <- 23 * cola - col
    row <- nrow(data$stats)
    
    cat(format(paste(data$yvar, "by", data$xvar), width = ow, justify = "centre"), 
        "\n")
    cat(rep("-", ow), sep = "", "\n")
    cat("|")
    for (i in seq_len(cola)) {
        cat(format(data$stats[1, i], width = n, justify = "right"), "|", sep = "")
    }
    cat("\n")
    cat(rep("-", ow), sep = "", "\n")
    for (i in 2:row) {
        cat("|")
        for (j in seq_len(cola)) {
            cat(format(data$stats[i, j], width = n, justify = "right"), "|", sep = "")
        }
        cat("\n")
    }
    cat(rep("-", ow), sep = "", "\n")
    
}


print_screen <- function(x) {
    
    columns <- c("  Column Name  ", "  Data Type  ", "  Levels  ", "  Missing  ", 
        "  Missing (%)  ")
    len_col <- as.vector(sapply(columns, nchar))
    xlev <- lapply(k$levels, paste, collapse = " ") %>% lapply(nchar) %>% unlist %>% 
        max
    lengths <- list(x$Variables, x$Types, xlev, x$Missing, x$MissingPer)
    n <- length(columns)
    nlist <- list()
    for (i in seq_len(n)) {
        nlist[[i]] <- max(len_col[i], max(sapply(lengths[[i]], nchar)))
    }
    clengths <- unlist(nlist)
    clengths[3] <- max(10, xlev)
    dash <- sum(clengths) + 6
    cat(rep("-", dash), sep = "")
    cat("\n|")
    for (i in seq_len(n)) {
        cat(columns[i], "|", sep = "")
    }
    cat("\n", rep("-", dash), sep = "")
    cat("\n")
    for (i in seq_len(x$Columns)) {
        cat("|", format(x$Variables[i], width = clengths[1], justify = "centre"), 
            "|", format(x$Types[i], width = clengths[2], justify = "centre"), "|", 
            format(paste(x$levels[[i]], collapse = " "), width = clengths[3], justify = "centre"), 
            "|", format(as.character(x$Missing[i]), width = clengths[4], justify = "centre"), 
            "|", format(as.character(x$MissingPer[i]), width = clengths[5], justify = "centre"), 
            "|\n", sep = "")
    }
    cat(rep("-", dash), sep = "")
    cat("\n\n")
    cat(" Overall Missing Values          ", x$MissingTotal, "\n", "Percentage of Missing Values    ", 
        x$MissingTotPer, "%\n", "Rows with Missing Values        ", x$MissingRows, 
        "\n", "Columns With Missing Values     ", x$MissingCols, "\n")
    
}
